### PR TITLE
Update to rust-bitcoin v0.19.1, add encode module

### PR DESCRIPTION
**Pull request by stevenroose**

### CODE REVIEW

File: src/encode.rs [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        }
+    };
+}
+impl_upstream!(u8);
+impl_upstream!(u32);
+impl_upstream!(u64);
+impl_upstream!([u8; 32]);
+impl_upstream!(Vec<u8>);
+impl_upstream!(Vec<Vec<u8>>);
+impl_upstream!(btcenc::VarInt);
```
Comment (apoelstra) [inline src/encode.rs]: I think we should `pub use bitcoin::VarInt` in `lib.rs`with_capacity

File: src/encode.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        }
+
+        impl Decodable for Vec<$type> {
+            #[inline]
+            fn consensus_decode<D: io::Read>(mut d: D) -> Result<Self, Error> {
+                let len = btcenc::VarInt::consensus_decode(&mut d)?.0;
+                let byte_size = (len as usize)
+                    .checked_mul(mem::size_of::<$type>())
+                    .ok_or(self::Error::ParseFailed("Invalid length"))?;
+                if byte_size > btcenc::MAX_VEC_SIZE {
```
Comment (apoelstra) [inline src/encode.rs]: and we should `pub use bitcoin::consensus::encode::MAX_VEC_SIZE` in this module

File: src/encode.rs (hunk 1 above)
Comment (stevenroose) [reply src/encode.rs]: Why? How about `Script`?

File: src/encode.rs (hunk 2 above)
Comment (stevenroose) [reply src/encode.rs]: Ah yeah instead of redefine. Agree with that.

File: src/encode.rs (hunk 1 above)
Comment (apoelstra) [reply src/encode.rs]: An elements script is not the same as bitcoin script :). But an elements varint is a bitcoin varint.

Comment (apoelstra) [reply src/encode.rs]: At some point we should have a separate `elements::script` and `elements::opcodes` modules that are independent of the Bitcoin ones, but for now (especially with our heavy Miniscript usage which is Bitcoin-only) it's fine to keep using `bitcoin::Script` in elements data structures, even though this isn't very well typed.

File: src/encode.rs [hunk 3]
```diff
@@ -22,6 +22,9 @@ use ::bitcoin::consensus::encode as btcenc;
 
 use transaction::{Transaction, TxIn, TxOut};
 
+pub use ::bitcoin::consensus::encode::MAX_VEC_SIZE;
```
Comment (apoelstra) [inline src/encode.rs]: Can we `pub use` this at the top level so that users can import `elements::VarInt`?

**Comment by apoelstra** [comment]
fuzztests are not finishing. I kicked the CI job. Hopefully this does not represent a perf loss. (It shouldn't; the new encoding code should have much *better* performance.)

**Comment by jonasnick** [comment]
Hmm travis still not happy

File: src/encode.rs (hunk 3 above)
Comment (stevenroose) [reply src/encode.rs]: Fixed this.

**Comment by stevenroose** [comment]
I'm confused with Travis, can't really see what changed except the upstream encoding logic. But like Andrew says, it should be more efficient not less. I had to force push, let's see if it changes anything.

**Comment by apoelstra** [comment]
I can reproduce the massive slowdown (like 20x) in the `deserialize_block` fuzzer locally. But it's hard to track something like this down.

**Comment by apoelstra** [comment]
I think the issue is that we now have a bunch more variable-sized objects in the block headers.

I recommend just dropping the number of fuzz iterations in travis by a factor 10.

**Pull request merged**